#412: Bitmap commands: Fixed deviation from redis implementation

apoorvyadav1111 opened the pull request:
### The changes that are included in the PR.
Addresses https://github.com/DiceDB/dice/issues/412

Fixed bitmap related commands to work with strings and strings + bytearrays

Design document, if any.

### Information on any implementation choices that were made.

Removed the optimization of resizing byte arrays in case the LSB is 0. This resulted in a deviation where a bit is set to 1 and then set back to 0. If the key to which this is set is the longest, the bitop commands return inaccurate results.

Additionally, GETSET was failing as we now handle returning byte arrays via GET. It turns out that the encoding constants are being reused even if they are different. In this case, the encoding value for deque and bytearray was 4. I have reshuffled the file contents for improved readability and possible changes in the values of the encoding to avoid collision in the future

### Evidence of sufficient testing. You MUST indicate the tests done, either manually or automated.
existing test cases pass + added new test cases to check this behaviour.

apoorvyadav1111 commented:
Hi @JyotinderSingh,

I have added some changes to the bit-related commands for handling strings. I see that I have some conflicts with the master branch, which is due to a PR addressing the same issues. I will resolve them by combining both (this branch handles more scenarios when it comes to resizing). Please review and provide feedback. I will continue finding more commands that need to handle values with their encoding as byte arrays.

Thanks,
Apoorv

apoorvyadav1111 commented:
Status Update:

Hi @JyotinderSingh, While rebasing with the master, I found a deviation in behavior in handling "int" data. Working on them as well in this PR.

JyotinderSingh commented:
Please rebase the PR, we just had a large change go in which might've introduced conflicts.

apoorvyadav1111 commented:
Hi @JyotinderSingh, I Updated the PR with a rebase. Please review. All tests passed, and I added new tests with a combination of encodings for bit commands on strings, byte arrays, and integers.

apoorvyadav1111 commented:
Hi @JyotinderSingh, and @soumya-codes, I have updated the code again after the second refactor. Please review and merge the PR if all checks out.

Kind regards,
Apoorv

JyotinderSingh reviewed (approved):
Thanks for these changes @apoorvyadav1111. The changes look good and well tested.
